Use RC interpolation over average in INSFV conservation tests

GiudGiud opened the pull request:
This time adding the failing recipes. 
PR introducing the issue: #16767

GiudGiud committed:
Move tests to Rhie Chow interpolation to avoid diffs due to pressure oscillations
Exact conservation is expected, but not reported in postprocessor outputs due to #16585
Refs #16918

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/676629/) on 00bac3d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16917/site)

This comment will be updated on new commits.

GiudGiud commented:
fv_quad_xy_noslip is failing with RC interpolation, with 2 MPI and 2 threads in debug mode only. This is a pretty specific configuration for failure, but I'm able to reproduce it consistently on my machine. Investigating

loganharbour commented:
By failing - you just mean diffing, right?

GiudGiud commented:
yes

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/conservation_INSFV.i:
```diff
@@ -245,9 +245,8 @@ velocity_interp_method='average'
-  # preconditioner chosen for repeatability, for performance use asm or others
-  petsc_options_iname = '-pc_type -pc_factor_shift_type'
-  petsc_options_value = 'lu       NONZERO'
+  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_pc_type -sub_pc_factor_shift_type'
+  petsc_options_value = 'asm      200                lu           NONZERO'
```
Why?

GiudGiud replied:
Going back from LU because it wasnt converging for average interpolation.

GiudGiud replied:
but this is a separate issue. We re having a diff with rc, with lu and asm

GiudGiud replied:
https://civet.inl.gov/job/676718/ for asm diff

lindsayad replied:
You should check the condition number for this problem. `-pc_type svd pc_svd_monitor` (make sure problem is < 1000 dofs)

GiudGiud replied:
No threads, no MPI
 0 Nonlinear |R| = 1.141567e+02
      SVD: condition number 6.259959741131e+04, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.529726220628e-04 7.042123243808e-04 1.272279687127e-03 1.612161798012e-03 1.679989449127e-03
      SVD: largest singular values : 8.907091144641e+00 8.920760075913e+00 9.214173027458e+00 9.447507543685e+00 9.576024556084e+00
      0 Linear |R| = 1.141567e+02
      1 Linear |R| = 1.189965e-11
 1 Nonlinear |R| = 3.038182e+01
      SVD: condition number 4.082403300291e+06, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 2.521446755486e-05 3.660251616165e-05 4.374527439737e-05 4.925648219373e-05 5.174696168739e-05
      SVD: largest singular values : 8.847327873058e+01 9.214111913125e+01 9.405357229165e+01 9.788559335175e+01 1.029356255610e+02
      0 Linear |R| = 3.038182e+01
      1 Linear |R| = 2.357727e-11
 2 Nonlinear |R| = 6.788610e+00
      SVD: condition number 2.240412492980e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 4.602093497856e-07 1.098516887710e-05 1.196363709071e-05 1.678841488737e-05 2.123108766478e-05
      SVD: largest singular values : 9.092693187950e+01 9.408647464700e+01 9.578191374576e+01 9.856647151483e+01 1.031058776646e+02
      0 Linear |R| = 6.788610e+00
      1 Linear |R| = 1.255511e-11
 3 Nonlinear |R| = 1.521793e-01
      SVD: condition number 9.171708180006e+07, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.201618549885e-06 9.180321367420e-06 1.009032971830e-05 1.508152464353e-05 1.767381544779e-05
      SVD: largest singular values : 1.003534266828e+02 1.022783714585e+02 1.051165889272e+02 1.091314094551e+02 1.102089468323e+02
      0 Linear |R| = 1.521793e-01
      1 Linear |R| = 3.234128e-12
 4 Nonlinear |R| = 2.840345e-05
      SVD: condition number 1.078669411658e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.042166334648e-06 9.060481628959e-06 9.590745257623e-06 1.311119275375e-05 1.706163152425e-05
      SVD: largest singular values : 1.014002099143e+02 1.045334866841e+02 1.056234993070e+02 1.112625084934e+02 1.124152947045e+02
      0 Linear |R| = 2.840345e-05
      1 Linear |R| = 6.307550e-16
 5 Nonlinear |R| = 1.088787e-13

2 threads, 2 mpi, debug mode (diffing configuration):

 0 Nonlinear |R| = 1.141567e+02
      SVD: condition number 6.259959741135e+04, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.529726220627e-04 7.042123243806e-04 1.272279687127e-03 1.612161798012e-03 1.679989449127e-03
      SVD: largest singular values : 8.907091144641e+00 8.920760075913e+00 9.214173027458e+00 9.447507543685e+00 9.576024556084e+00
      0 Linear |R| = 1.141567e+02
      1 Linear |R| = 1.283775e-11
 1 Nonlinear |R| = 3.038182e+01
      SVD: condition number 4.082403300254e+06, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 2.521446755509e-05 3.660251616180e-05 4.374527439591e-05 4.925648219373e-05 5.174696168739e-05
      SVD: largest singular values : 8.847327873058e+01 9.214111913125e+01 9.405357229165e+01 9.788559335175e+01 1.029356255610e+02
      0 Linear |R| = 3.038182e+01
      1 Linear |R| = 1.886399e-11
 2 Nonlinear |R| = 6.788610e+00
      SVD: condition number 2.240412494214e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 4.602093495315e-07 1.098516887688e-05 1.196363709003e-05 1.678841488750e-05 2.123108766343e-05
      SVD: largest singular values : 9.092693187953e+01 9.408647464694e+01 9.578191374568e+01 9.856647151472e+01 1.031058776644e+02
      0 Linear |R| = 6.788610e+00
      1 Linear |R| = 1.299889e-11
 3 Nonlinear |R| = 1.521793e-01
      SVD: condition number 9.171708176712e+07, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.201618550312e-06 9.180321366604e-06 1.009032971955e-05 1.508152464616e-05 1.767381544796e-05
      SVD: largest singular values : 1.003534266825e+02 1.022783714579e+02 1.051165889270e+02 1.091314094535e+02 1.102089468318e+02
      0 Linear |R| = 1.521793e-01
      1 Linear |R| = 3.143176e-12
 4 Nonlinear |R| = 2.840347e-05
      SVD: condition number 1.078669411664e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 1.042166334633e-06 9.060481629277e-06 9.590745257613e-06 1.311119275427e-05 1.706163152454e-05
      SVD: largest singular values : 1.014002099140e+02 1.045334866832e+02 1.056234993069e+02 1.112625084931e+02 1.124152947035e+02
      0 Linear |R| = 2.840347e-05
      1 Linear |R| = 6.185062e-16
 5 Nonlinear |R| = 1.369528e-13
 Solve Converged!

GiudGiud replied:
oh actually that's not for the failing test

GiudGiud replied:
For the failing test in the failing configuration:

 0 Nonlinear |R| = 1.145739e+02
      SVD: condition number 2.346729017364e+03, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 4.081557487698e-03 8.004931259928e-03 9.698643282531e-03 1.194772475952e-02 1.293113889030e-02
      SVD: largest singular values : 9.154029694036e+00 9.215824274390e+00 9.223761592282e+00 9.447480604893e+00 9.578309392419e+00
      0 Linear |R| = 1.145739e+02
      1 Linear |R| = 1.407774e-11
 1 Nonlinear |R| = 1.300242e+01
      SVD: condition number 1.245388384970e+07, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 8.296914646474e-06 1.679709477761e-05 3.001276497910e-05 4.084153876655e-05 4.406878345211e-05
      SVD: largest singular values : 8.911197053621e+01 9.268885617389e+01 9.451606274255e+01 9.818502187516e+01 1.033288113181e+02
      0 Linear |R| = 1.300242e+01
      1 Linear |R| = 2.997326e-11
 2 Nonlinear |R| = 8.234543e-01
      SVD: condition number 1.520626475665e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 7.004766506755e-07 1.400188218740e-06 2.110794867808e-06 4.025210194633e-06 4.697501994878e-06
      SVD: largest singular values : 9.876713811227e+01 1.013171703966e+02 1.030611746385e+02 1.054304661127e+02 1.065163340602e+02
      0 Linear |R| = 8.234543e-01
      1 Linear |R| = 5.782767e-12
 3 Nonlinear |R| = 3.005242e-04
      SVD: condition number 2.600920449959e+08, 0 of 448 singular values are (nearly) zero
      SVD: smallest singular values: 4.542254551557e-07 9.279341519870e-07 1.409518131510e-06 2.613073598344e-06 3.309658270434e-06
      SVD: largest singular values : 1.047274223774e+02 1.066677317940e+02 1.099629822521e+02 1.155444311260e+02 1.181404275207e+02
      0 Linear |R| = 3.005242e-04
      1 Linear |R| = 2.684896e-15
 4 Nonlinear |R| = 2.016780e-13
 Solve Converged!

GiudGiud replied:
Seems fairly similar to me, and 1e8 condition number isnt great but it's not 1e-4 diff bad.

lindsayad replied:
1e8 is higher than I would like. I turned on automatic scaling and `scaling_group_variables='u v'`, but that only reduces it to 4e7. Frustrating that it is so high even with close to unity material properties and mesh size. Motivation to me to pursue for exotic automatic scaling schemes.

Anyway, even though the conditioning is not what I want it to be, the solve still seems to go fine. If I run an exodiff on the serial vs. distributed-parallel-threaded case, the only thing that diffs is that darn `mid2_advected_energy`

GiudGiud replied:
So the mesh values are the same but the postprocessor is not?
This point to a difference in cached internal face values imo

lindsayad replied:
> So the mesh values are the same but the postprocessor is not?

correct

> This point to a difference in cached internal face values imo

I concur

lindsayad reviewed (dismissed):
Let's do a full stop here. If we're getting csv-diffs with LU then there truly is a bug somewhere in the code as opposed to simple lack of repeatability due to different ordering of parallel communication. With LU we should be repeatable, so if we're not then that means we're not correctly implementing thread-join or something else

lindsayad commented:
Oh I see that solves are actually failing...

lindsayad commented:
Yea and i see that the solves are failing after #16911. Well at least we have the targets added to this PR

lindsayad committed:
Implement computeElemValuesFace and computeNeighborValuesFace for FV

We were getting incorrect results in our interface postprocessor because
the FV solutions were not initialized

lindsayad committed:
Pass current_elem_is_fi_elem to Moose::FV::interpolate

lindsayad committed:
Regold after interface value calculation framework fix

loganharbour commented:
Since you can't approve your own PR 😆

lindsayad reviewed (commented):
Removing my requested changes...

loganharbour commented:
Forgot about that darn skipped test...

lindsayad committed:
nonsingular test should pass with SOME pc

lindsayad commented:
So close :cry:
